pennylane/data/attributes/dictionary.py

```diff
@@ -0,0 +1,92 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains an AttributeType that allows for heterogenous dictionaries
+of Dataset attributes."""
+
+
+import typing
+from collections.abc import Mapping, MutableMapping
+from typing import Dict, Generic, Optional, TypeVar, Union
+
+from pennylane.data.base.attribute import AttributeType
+from pennylane.data.base.mapper import MapperMixin
+from pennylane.data.base.typing_util import HDF5Any, HDF5Group, T
+
+
+class DatasetDict(
+    Generic[T],
+    AttributeType[HDF5Group, typing.Mapping[str, T], Optional[typing.Mapping[str, T]]],
+    MutableMapping,
+    MapperMixin,
+):
```

Comment:
Can we implement a check that the keys of the dictionary are strings? Otherwise, the error is not obvious:
```python
>>> import pennylane as qml
>>> x = qml.data.attributes.DatasetDict({1: 2, 3: 4})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mikhail/ro/pennylane/pennylane/data/base/attribute.py", line 248, in __init__
    self.__post_init__(value, info)
  File "/Users/mikhail/ro/pennylane/pennylane/data/attributes/dictionary.py", line 41, in __post_init__
    self.update(value)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/_collections_abc.py", line 941, in update
    self[key] = other[key]
  File "/Users/mikhail/ro/pennylane/pennylane/data/attributes/dictionary.py", line 66, in __setitem__
    self._mapper[__key] = __value
  File "/Users/mikhail/ro/pennylane/pennylane/data/base/mapper.py", line 117, in __setitem__
    self.set_item(key, value, None)
  File "/Users/mikhail/ro/pennylane/pennylane/data/base/mapper.py", line 112, in set_item
    attr_type(value, info, parent_and_key=(self.bind, key))
  File "/Users/mikhail/ro/pennylane/pennylane/data/base/attribute.py", line 246, in __init__
    self._bind = self._set_value(value, info, parent, key)
  File "/Users/mikhail/ro/pennylane/pennylane/data/base/attribute.py", line 315, in _set_value
    new_bind = self.value_to_hdf5(parent, key, value)
  File "/Users/mikhail/ro/pennylane/pennylane/data/attributes/scalar.py", line 33, in value_to_hdf5
    bind_parent[key] = value
  File "/Users/mikhail/ro/pennylane/venv/lib/python3.9/site-packages/h5py/_hl/group.py", line 465, in __setitem__
    name, lcpl = self._e(name, lcpl=True)
  File "/Users/mikhail/ro/pennylane/venv/lib/python3.9/site-packages/h5py/_hl/base.py", line 200, in _e
    name = name.encode('ascii')
AttributeError: 'int' object has no attribute 'encode'
```

Reply:
done!